### PR TITLE
MC/diffusive hybrid integration

### PR DESCRIPTION
**Pull request by awlostowski-noaa**
This PR integrates the MC/diffusive hybrid routing capability. With these additions, users can run simulations that run MC on part of the domain, and diffusive wave (CN-mod) on another. This represents a big step towards a coastal-coupled inland hydraulics model, because now we can run diffusive routing in coastal adjacent domains. The focus of this PR is engineering the hybrid routing scheme - additional work is needed to assure solution accuracy. Moreover, this represents a first-cut of the hybrid routing scheme, surely more work will be needed regarding parallelization, water body integration and data assimilation. Additional modifications will follow, as we gain clarity on what exactly the final product needs to do and where it needs to do it. 

To run a hybrid simulation, the user must include a new parameter set in the configuration (`.yaml`) file under `compute_parameters` called `hybrid_parameters`. Under `hybrid_parameters` the variable `diffusive_domain` points to a `.yaml` file that stores a dictionary of diffusive domain segment IDs, organized by tailwater. Below is an example snippet of a configuration file for a hybrid simulation:

``` yaml
compute_parameters:
    parallel_compute_method: serial
    compute_kernel: V02-structured
    hybrid_parameters:                            # <-- new sub-section for hybrid parameters
        diffusive_domain: diffusive_domain.yaml   # <-- dictionary of diffusive domain segments 
    forcing_parameters:
        qts_subdivisions: 12
        dt: 300
        qlat_input_folder: /some/file/directory/
        qlat_file_pattern_filter: "*.CHRTOUT_DOMAIN1"
        qlat_file_index_col: feature_id
        qlat_file_value_col: q_lateral
        nts: 288
        max_loop_size: 24
```
As mentioned above, the `diffusive_domain` parameter points to a `.yaml` file containing a dictionary of diffusive domain segments. Dictionary keys are the tailwater segments of the diffusive sub-networks, dictionary values area  list of segment IDs in a diffusive sub-domain. Here is an example of a `diffusive_domain` file for single sub-domain at the mouth of the Cape Fear River:
``` yaml
8835404:  # <--- This is the network tailwater segment ID
    - 8832730  # <--- This is a list of segment IDs in the network, including the tailwater
    - 8832674
    - 8832666
    - 8832664
    - 8832654
    - 8832656
    - 8832652
    - ...
```

There are several key limitations of the MC/diffusive hybrid model at this time:

- Diffusive sub-domains must extend all the way to network tailwaters. MC domain simulations can only provide inflows to the diffusive wave domain. At this time, diffusive wave domain simulations cannot provide inflows to the MC domain. We will add this capability as needed, once the full extend of the CONUS diffusive wave domain becomes clear.
- No parallel scheme for the diffusive simulations, yet. However, the code structure in `compute.compute_diffusive_routing` is predisposed for by-network parallelization.
- No data assimilation. We are holding off on this and reserving it for a forthcoming PR. 
- No reservoirs. Unclear as to whether or not this is needed, so holding off until the full diffusive domain extent becomes clearer. 

## Additions

- New variables are read in from the configuration file, under `compute_parameters`
- `preprocess.nwm_network_preprocess` returns a dictionary called `diffusive_network_data` that packages network data for each diffusive sub-network. 
- If a `diffusive_domain` variable is provided, then all returns from `preprocess.nwm_network_preprocess` will be derived for non-diffusive network segments only. This way, MC routing is only conducted on segments where diffusive routing will not also be run. Tributary segments adjacent to the diffusive domain(s) become tailwater segments in the MC domain.
- `__main__.nwm_route` is edited to include a new call to `compute.compute_diffusive_routing` for hybrid simulations. This call comes after the existing call to `compute.compute_nhd_routing_v02`, which is used to execute the MC calculations. 
- A new function is added to `compute.py` called `compute_diffusive_routing`. This function is responsible for building inputs to and calling the diffusive routing algorithm.
- `diffusive_utils.py` was heavily revised to build inputs for the hybrid version of the diffusive wave code and utilize DataFrame inputs instead of strictly numpy array inputs. Now that this script is called from python instead of cython, we can be more flexible with the input structures. 
- `diffusive.f90` (CN-mod) was heavily revised. First and foremost, the code is now bespoke for mainstem domains, where off-network inflows are handled through junction boundaries. Second, the code has been re-formatted and some unused variable have been removed. More formatting and cleaning is still needed by this is a start. 

## Removals

- All other use cases of diffusive wave code have been removed. The only use-case at this time is the MC/diffusive hybrid with the CN-mod code. 

## Testing

1. /glade/u/home/adamw/projects/t-route/test/jobs/hybrid_integration/Cape_Fear_hybrid_test.yaml

## Screenshots
Below is a plot showing a purely MC simulation (BLUE) and a MC/diffusive hybrid simulation from a stream segment in Wilmington North Carolina, near the mouth of the Cape Fear river. The simulation time domain was ~23 days and included the major event of Hurricane Florence. We can see that the diffusive hybrid model produces a numerically stable solution, that differs quite a bit from the MC, in terms of peak flow timing/magnitude. 
image

Both MC and diffusive components of this hybrid simulation were run in serial. The MC routing code took 184 seconds to execute, while the diffusive code took 74 seconds to execute. The diffusive domain is only 43 segments, while the MC domain contains 14861 segments. So, the MC code took ~0.01 seconds per segment and the diffusive code took ~1.7 seconds per segment. This timing disparity is partially due to the very high flows being simulated. Because of the adaptive time stepping, the diffusive model runs with extremely short timesteps to maintain numerical stability. Shorter simulation timesteps increase the wall clock time. 

## Todos

- Additional QA/QC of diffusive Fortran code to ensure that mass balance is being sustained throughout simulation.
- Additional formatting and cleaning of the diffusive Fortran code
- Consider schemes to modularize the diffusive Fortran kernel to the level of a single reach (sort of like the MC kernel), this would allow us greater flexibility in implementing various computations schemes, integrating more seamlessly with MC solutions, and liming short timestep durations to only those reaches that most require it. 

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### CODE REVIEW

**Review by jhrehanoaa — COMMENTED** [review]
Suggestions from experimentations:

After pulling this apart and making some super simple test cases I can see some possible solutions. This can be found in "/glade/u/home/jhreha/t-route/test/jobs/parity_check/". I made a test case with 
3 MC ids: 
8810287 
8810285 - diffusive
8810283
8810289
8810281 - diffusive
and 
2 diffusive ids: 
8810285:
    - 8810281

For some network arrangement context:
(Pdb) rconn
`defaultdict(None, {8810283: [], 8810285: [8810283], 8810289: [8810285, 8810287], 8810287: []})`

(Pdb) diffusive_network_data[8810285]['rconn']
`defaultdict(None, {8810281: [], 8810285: [8810281]})`



- In main.py, line 1262  the output of the MC run will be finished. We can then bring the fvd back in for use with diffusive.  

Which will bring in the 4 MC output IDs. 
`(Pdb) pd.read_pickle("/glade/u/home/jhreha/t-route/test/jobs/parity_check/sim1/sim1_hybrid")
           (0, q)    (0, v)    (0, d)    (1, q)    (1, v)    (1, d)  ...    (3, q)    (3, v)    (3, d)    (4, q)    (4, v)    (4, d)
8810283  0.008463  0.145497  0.032826  0.008502  0.145753  0.032916  ...  0.008533  0.145952  0.032986  0.008539  0.145987  0.032998
8810285  0.043131  0.032202  0.689043  0.039729  0.026552  0.479617  ...  0.035262  0.025934  0.459158  0.033887  0.031063  0.643622
8810287  0.000351  0.040176  0.003940  0.000333  0.039284  0.003808  ...  0.000312  0.037760  0.003588  0.000305  0.037250  0.003515
8810289  0.063651  0.217682  0.100557  0.063590  0.217550  0.100459  ...  0.060725  0.213803  0.097692  0.058743  0.211162  0.095762`


8810285 is the ID that will need to be combined with upstream flows from MC. It takes input from ID 8810289 (MC). Since rconn and the rconn from diffusive_network_data have separate networks they no longer preserve the correct upstreams for each segment. I suggest we preserve a master rconn/conn or some conn object that contains both upstreams for an ID that are MC and diffusive. This can be kept in order to quickly pull the correct upstream ids when diffusive is being used. 

When the MC output finishes on line 1262 I imagine we can have some sort of repeat of the run iterator call that calls the compute_kernal.  `for run_set_iterator, run in enumerate(run_sets):` set off by a check for hybrid_parameters. The only question is inside of the mc_reach how can we update diffusive fvd with the MC upstreams in realtime? I figured we could pass the values the same way usgs overwrites values. Wasn't sure how we wanted to approach this part of the hybrid model since q0 only covers the first timestep so we can't use that. I imagine this is mc_reach splicing of some sort where we take the current timestep of diffusive = upstream MC + upstream diffusive. We can talk through this further as this mc_reach aspect will likely be the largest hurdle. Just wanted to share some thoughts and think through the next steps more with you.

**Review by jhrehanoaa — APPROVED** [review]
test case `/glade/u/home/adamw/projects/t-route/test/jobs/hybrid_integration/Cape_Fear_hybrid_test.yaml` works as expected. I reconfigured this test case to run classic diffusive and in serial/parallel. Additionally, ran florence test cases as well. All runs completed successfully. 

 When running one of the parity check files (sim1.yaml) from last week I am getting concatenation problems with results. 
`  File "/glade/u/home/jhreha/miniconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/glade/u/home/jhreha/miniconda3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/glade/u/home/jhreha/t-route/src/nwm_routing/src/nwm_routing/__main__.py", line 1677, in <module>
    main_v03(v_args[1])
  File "/glade/u/home/jhreha/t-route/src/nwm_routing/src/nwm_routing/__main__.py", line 1226, in main_v03
    q0 = new_nwm_q0(run_results)
  File "/glade/u/home/jhreha/t-route/src/nwm_routing/src/nwm_routing/__main__.py", line 959, in new_nwm_q0
    return pd.concat(
  File "/glade/u/home/jhreha/miniconda3/lib/python3.8/site-packages/pandas/core/reshape/concat.py", line 285, in concat
    op = _Concatenator(
  File "/glade/u/home/jhreha/miniconda3/lib/python3.8/site-packages/pandas/core/reshape/concat.py", line 342, in __init__
    raise ValueError("No objects to concatenate")
ValueError: No objects to concatenate
`
I would double check yours still works. Other than that it looks to be operating as designed.

**Review by kumdonoaa — APPROVED** [review]
Fortran looks clean with JULES format. Thanks for correcting 1 by mstem_frj(1).  The diffusive_utils file looks good with simply finding geo data by param_df.  Thanks for getting this heavy lift done.

File: src/nwm_routing/src/nwm_routing/preprocess.py [hunk 1]
```diff
@@ -123,8 +195,8 @@ def nwm_network_preprocess(
     if break_network_at_waterbodies:
         network_break_segments = network_break_segments.union(wbody_conn.values())
     if break_network_at_gages:
-        network_break_segments = network_break_segments.union(gages.keys())
-
+        network_break_segments = network_break_segments.union(gages['gages'].keys())
+
```
Comment (awlostowski-noaa) [inline src/nwm_routing/src/nwm_routing/preprocess.py:199]: This is a bug fix. The `gages` dictionary object has a first-level key of `gages`. Below this, the keys are linkIDs and the values are gageIDs. Previously, we were just adding a string, `'gages'` to the set `network_break_segments`, and as a result gage segments were being left out.

**Pull request merged**